pkg/services/alerting/notifiers/slack.go

```diff
@@ -305,28 +291,88 @@ func (sn *SlackNotifier) Notify(evalContext *alerting.EvalContext) error {
 		return err
 	}
 
-	cmd := &models.SendWebhookSync{
-		Url:        sn.URL,
-		Body:       string(data),
-		HttpMethod: http.MethodPost,
+	if err := sn.sendRequest(evalContext.Ctx, data); err != nil {
+		return err
+	}
+
+	if sn.Token != "" && sn.UploadImage {
+		err := sn.slackFileUpload(evalContext, sn.log, "https://slack.com/api/files.upload", sn.Recipient, sn.Token)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (sn *SlackNotifier) sendRequest(ctx context.Context, data []byte) error {
+	const url = "https://slack.com/api/chat.postMessage"
+	sn.log.Debug("Sending Slack API request", "url", url, "data", string(data))
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(data))
+	if err != nil {
+		return fmt.Errorf("failed to create HTTP request: %w", err)
 	}
+
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Set("User-Agent", "Grafana")
 	if sn.Token != "" {
 		sn.log.Debug("Adding authorization header to HTTP request")
-		cmd.HttpHeader = map[string]string{
-			"Authorization": fmt.Sprintf("Bearer %s", sn.Token),
-		}
+		request.Header.Set("Authorization", fmt.Sprintf("Bearer %s", sn.Token))
 	}
-	if err := bus.DispatchCtx(evalContext.Ctx, cmd); err != nil {
-		sn.log.Error("Failed to send slack notification", "error", err, "webhook", sn.Name)
+
+	netTransport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			Renegotiation: tls.RenegotiateFreelyAsClient,
+		},
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout: 30 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 5 * time.Second,
+	}
+	netClient := &http.Client{
+		Timeout:   time.Second * 30,
+		Transport: netTransport,
+	}
+	resp, err := netClient.Do(request)
+	if err != nil {
 		return err
 	}
-	if sn.Token != "" && sn.UploadImage {
-		err = sn.slackFileUpload(evalContext, sn.log, "https://slack.com/api/files.upload", sn.Recipient, sn.Token)
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			sn.log.Warn("Failed to close response body", "err", err)
+		}
+	}()
+
+	if resp.StatusCode/100 == 2 {
```

Comment:
I like this, so we can handle any type of 2xx statuses, but which other 2xx errors have you seen apart from 200?
204 maybe? (a question really - not a comment)

Reply:
I haven't seen other statuses personally, but this check was written before my time I think, and in any case it's best practice to support all 2xx statuses. Even if they haven't happened so far, they could start happening.